src/jltypes.c

```diff
@@ -1833,13 +1833,43 @@ static unsigned typekeyvalue_hash(jl_typename_t *tn, jl_value_t *key1, jl_value_
     return hash ? hash : 1;
 }
 
+// set the `has_concrete_subtype` flag of `dt` to zero for type parameters
+// that are illegal for certain types
+int jl_precompute_has_concrete_subtype(jl_datatype_t *dt) {
+    if (dt->name == jl_genericmemory_typename) {
+        jl_value_t *eltype = jl_tparam1(dt);
+        if (!jl_is_type(eltype) && !jl_is_typevar(eltype))
+            return 0;
+    }
+    if (dt->name == jl_tuple_typename) {
+        size_t i, l = jl_nparams(dt);
+        for (i = 0; i < l; i++) {
+            jl_value_t *p = jl_tparam(dt, i);
+            if (jl_is_vararg(p))
+                p = ((jl_vararg_t*)p)->T;
+            if ((p && !jl_is_type(p) && !jl_is_typevar(p)) || p == jl_bottom_type)
+                // tuple types like Tuple{:x} and Tuple{Union{}} cannot have instances
+                return 0;
+        }
```

Comment:
IIUC, with a bit more refactoring, you can share the codepath `jl_compute_has_concrete_subtype_from_fields` for these too?